### ISSUE TITLE
Remove the margin bottom from popups

### ISSUE THREAD

**Issue by predragnikolic** [sublimelsp/LSP#1553]
**Describe the bug**
There is a noticiably bigger margin bottom at bottom of the popups.

Current behavior:
![image](https://user-images.githubusercontent.com/22029477/105580105-bf393700-5d8a-11eb-97ce-df1815530da0.png)

Expected behavior:
![image](https://user-images.githubusercontent.com/22029477/105580117-c7917200-5d8a-11eb-8b35-4aed2de92467.png)

This is due to ST not supporting margin collapsing. - https://github.com/sublimehq/sublime_text/issues/3828

While this is a ST bug, 
LSP used to show the expected popup behavior before d772f935bc062a5e7a0c5443576cbdeb246c8188 was merged.
![image](https://user-images.githubusercontent.com/22029477/105580117-c7917200-5d8a-11eb-8b35-4aed2de92467.png)

So I wanted to ask if it would be ok that we only revert the margin, padding changes(the font change would stay) introduced in d772f935bc062a5e7a0c5443576cbdeb246c8188 until ST fixes margin collapsing?

**Comment by rchl** [comment]
Wouldn't that regress some cases? I mean, the pop-up content can vary and we should consider all cases when changing margins. For example the case when action links are enabled.

I'd like us to figure out a solution that fixes the issue for all cases and at the same time doesn't regress when the actual bug is fixed. That would probably require adding extra elements just for the purpose of making margins work in all cases.

**Comment by predragnikolic** [comment]
Based on what you discovered:
> BTW. ST seems to handle margin-collapsing properly for adjacent elements but not in the case of parent and child elements.

What if we applied margin-bottom only for the first children of the `.lsp_popup` class?  Like `.lsp_popup > p` in css. 
ST doesn't support that syntax (`.lsp_popup > p`), but we can archive the same with the following:
 
```diff
diff --git a/popups.css b/popups.css
index 5738bd0..11476af 100644
--- a/popups.css
+++ b/popups.css
@@ -1,7 +1,15 @@
 .lsp_popup {
-    margin: 0.5rem;
+    margin: 0.5rem 0.5rem 0 0.5rem;
     font-family: system;
 }
+/* this SECTION is the same as .lsp_popup > p { margin-bottom: 0.5rem; }*/
+.lsp_popup p {
+    margin-bottom: 0.5rem;
+}
+.lsp_popup p p {
+    margin-bottom: 0;
+}
+/* END SECTION */
 .lsp_popup hr {
   border-color: color(var(--foreground) alpha(0.10));
 }
@@ -38,6 +46,7 @@
     padding: 0.5rem;
 }
 .actions {
+    margin-bottom: 0.5rem;
     font-family: system;
     border-width: 0;
     background-color: color(var(--foreground) alpha(0.1));

```

The above popup.css file, initially seem to fixes all the problems I have with the current popup (with enabled/ or disabled `show_symbol_action_links`) but will have to test more :)

**Comment by rchl** [comment]
```
+/* this SECTION is the same as .lsp_popup > p { margin-bottom: 0.5rem; }*/
+.lsp_popup p {
+    margin-bottom: 0.5rem;
+}
+.lsp_popup p p {
+    margin-bottom: 0;
+}
```

Not entirely the same as `>` since the `.lsp_popup p p` can end up overriding some other selectors like `.mdpopups p` (made it up, not sure if it actually exists).

**Comment by predragnikolic** [comment]
The `.mdpopups p` selector does exist. But `.lsp_popup p p` will only override `p` that are inside of `p`.  I just now realized that this doesn't make sense semantically. The `p` tag shouldn't  allow other block level elements inside it. (see https://stackoverflow.com/questions/8397852/why-cant-the-p-tag-contain-a-div-tag-inside-it).

**Comment by rchl** [comment]
> But `.lsp_popup p p` will only override `p` that are inside of `p`.

It will also apply to the innermost element in:
```html
<div class="lsp_popup">
  <p>
    <div class="mdpopups">
      <p>Foo</p>
    </div>
   </p>
</div>

> I just now realized that this doesn't make sense semantically. The `p` tag shouldn't allow other block level elements inside it.

Yes, good point. In a real HTML parser that would not be allowed and nested `p`'s would be made into siblings. I don't think minihtml implements that part of the spec but we should still not create code like that.

**Comment by predragnikolic** [comment]
I do not think that this hierarchy is possible `lsp_popup > mdpopups` 
```
<div class="lsp_popup">
    <div class="mdpopups">
       ...popup content
    </div>
</div>
```
based on the docs I tought that hierarchy is `mdpopups > lsp_popup` 
```
<div class="mdpopups">
    <div class="lsp_popup">
       ...popup content
    </div>
</div>
```
https://facelessuser.github.io/sublime-markdown-popups/api/#styling
> This way plugins can inject CSS to style their specific elements via `.mdpopups .myplugin-wrapper .myclass {}` or simply `.myplugin-wrapper .myclass {}`.

**Comment by rchl** [comment]
I think we are focusing on the wrong element anyway because it's the `div.highlight` that causes the issue here.

I have some idea how to fix this.

**Comment by predragnikolic** [comment]
It is not nessessarily a div.highlight, for example the selector is code.highlight in case of LSP-json sometimes. In pyls case i belive, that server returns just a p tag. (I would have to check that)

**Comment by rwols** [comment]
Servers return markdown.

**Comment by rchl** [comment]
Attempted to fix it with #1568 but unfortunately, it's really hard to fix the issue without accidentally removing some wanted margins somewhere.

The extreme way to fix that would be to parse HTML with some HTML parsing library and apply modifications to margins from HTML but that would be really inefficient and overkill.
